### ISSUE TITLE
read_parquet() does not capture divisions when partitioned

### ISSUE THREAD

**Issue by gorlins** [dask/dask#2991]
Reading a dataframe from parquet fails to load the partitions in order or preserve an index, while it works if reading the same dataframe from a globstring

Expected behavior: partitions in the dask frame correspond to ordered partitions of the parquet file, and divisions on the index should be preserved if partitioning and order of partitions does not scramble the index (ie if the partition has the same ordering as the index, is a prefix, etc)

```python
from dask import dataframe as dask_df
import pandas as pd
from datetime import date, timedelta, datetime
import dask
print(dask.__version__) # 0.16.0
import fastparquet
print(fastparquet.__version__) . # 0.1.3

def make_dt(hour):
    return datetime(2017, 1, 1, 0, 0, 0) + timedelta(hours=hour)

def make_part(hours):
    return pd.DataFrame([
        dict(
            hour=h,
            dt=make_dt(h),
            date=make_dt(h).date().strftime('%Y-%m-%d')
        )
        for h in hours
    ]
    ).set_index('hour')

ddf = dask_df.from_pandas(make_part(range(480,520)),npartitions=1)

ddf.to_parquet('partition_play', write_index=True, partition_on=['date'])

back = dask_df.read_parquet('partition_play', index='hour') # hour is actually known as default index

# back will .compute() correctly, but does not know its divisions even though the parquet file has them in the min/max part of the header
# I think this is because the partitions are not stored in order in the _metadata file
back.known_divisions # False

# The glob string will read the partitions in order and preserves the index
back_glob = = dask_df.read_parquet('partition_play/date=*/*.parquet')
back_glob.known_divisions # True

```

If the real issue is just row groups in _metadata not being sorted, I believe this could be addressed by just reordering the partitions in read_parquet before checking for an index, but I'm not too familiar w/ how parquet should work in this case

**Comment by martindurant** [comment]
Quick note: there is a certain order to the partitions as referenced in the `_metadata` file. Whether these are the same order as with a glob will depend on what pandas' group-by returns during partitioning - I would think the groups are indeed in order. In which case, there is probably a simple fix for this issue.

**Comment by gorlins** [comment]
in my example above, they are not in order, which causes the issue I believe.  Here is what is returned by the parquet stats:

```python
{'distinct_count': {'dt': [None, None], 'hour': [None, None]},
 'max': {'dt': [numpy.datetime64('2017-01-22T15:00:00.000000000'),
   numpy.datetime64('2017-01-21T23:00:00.000000000')],
  'hour': [519, 503]},
 'min': {'dt': [numpy.datetime64('2017-01-22T00:00:00.000000000'),
   numpy.datetime64('2017-01-21T00:00:00.000000000')],
  'hour': [504, 480]},
 'null_count': {'dt': [0, 0], 'hour': [0, 0]}}
```

Because they are out of order, fastparquet.api.sorted_partitioned_columns does not think 'hour' is sorted and therefore this part does not pick it up:

```python
        minmax = fastparquet.api.sorted_partitioned_columns(pf)
        if index_col in minmax:  # False
```

**Comment by gorlins** [comment]
actually i'm not sure if this is a dask bug or a fastparquet bug - should the row groups have been written in order (is that dask or fastparquet's job?), should sorted_partitioned_columns return sorted columns even if the row groups have them out of order, but they are still sortable, or should dask be able to handle all cases when requested on read (if not by default)?

**Comment by martindurant** [comment]
Each partition of a dask dataframe will be passed through a group-by independently, and the ordering in the metadata will be deterministic: all the parts of the original partition 1, all the parts of the original partition 1... 
If you want, for example, all the rows where field=val1 to come first, you need to split your data on that column, e.g., by setting it as the index, which causes a data shuffle. I don't recall if you can use the index in partition_on, you may need to reset_index too, to make those values a normal column after the sort/shuffling step.

**Comment by gorlins** [comment]
@martindurant I'm not sure that addresses the issue - in my above example, there was only one partition originally (reading back gets two because partition_on split the output, but there is only 1 .parquet file in each partition folder in my example).  Unfortunately it's difficult to keep track of when we're discussing dask partitions vs parquet partitions or .parquet files!

I've seen what you mention for each part file w/in a parquet partition folder aligning with the partitions of the dask DF that was written.  However, the _metadata file records the parquet row groups from the partitions out of order - at least when I run it - and thus Dask cannot recreate the divisions.  In fact, it actually loads the partitions out of order; you can see the head is from the second date in my example:

```python
back = dask_df.read_parquet('partition_play', index='hour')
print(back.get_partition(0).head()) # Expected part0/df.head to be 2017-01-21
                      dt       date
hour                               
504  2017-01-22 00:00:00 2017-01-22
505  2017-01-22 01:00:00 2017-01-22
506  2017-01-22 02:00:00 2017-01-22
507  2017-01-22 03:00:00 2017-01-22
508  2017-01-22 04:00:00 2017-01-22

print(back.get_partition(1).head()) . # 2017-01-21 loaded after 2017-01-22!
                      dt       date
hour                               
480  2017-01-21 00:00:00 2017-01-21
481  2017-01-21 01:00:00 2017-01-21
482  2017-01-21 02:00:00 2017-01-21
483  2017-01-21 03:00:00 2017-01-21
484  2017-01-21 04:00:00 2017-01-21
```

  Obviously this will not work generically, but in cases where the column to partition_on shares the same ordering as the index - in my case w/ date vs hours elapsed, or if the partition is a prefix of a string in a sorted file, etc - the intuitive behavior IMHO is to load it back the partitions in order, and therefore preserve the same sorting order w/ the same index/divisions as when it was written (by dask or otherwise); it appears the correct metadata does exist to be able to do so, and it's just a matter of either storing the row groups in the correct order or re-ordering the partitions on read in a way that makes sense.

**Comment by martindurant** [comment]
Can you try the following change in fastparquet.writer.partition_on_columns:

```
    for key in gb.indices:
->
    for key in sorted(gb.indices):
```

**Comment by gorlins** [comment]
ah, yes that looks like it does the trick on the write side.  I'm noticing that my results are not 100% repeatable, which makes sense since gb.indices is a dict and the iteration order is not guaranteed.  The tests pass iff it iterates in order.

Does it make sense to add in functionality to reorder the row groups on the read side? ie if an index is explicitly requested (vs just inferred) then the row groups can be ordered by min value of that field and an index with divisions can be built if the data happens to be separated.  If the max/min overlap, the divisions can always be discarded as they are currently.

**Comment by martindurant** [comment]
Yes @gorlins , I suppose all of that information is available at the time of loading, and in general I can imagine a `fastparquet.ParquetFile.reorder_rowgroups(index)` method, which does a sort on the row_groups by the statistics min; dask can optionally call this. Note that in the future, index may be a tuple for multi-level indexing. Would you be interested in writing such a function? 
That, together with the small fix above, would make a nice PR.

**Comment by gorlins** [comment]
i could potentially write the new feature in a few weeks as I have a few upcoming deadlines... I would recommend putting the sorting in immediately as an independent task since it clearly improves the behavior

**Comment by mrocklin** [comment]
@martindurant was this resolved?  Can I ask you to either summarize what should happen in dask here or, if there are no concrete changes that should be made to the codebase, close?

**Comment by martindurant** [comment]
The change was already made on the fastparquet side, so this kind of unsorted dataset should not happen any more. Nevertheless, the idea of reordering the row-groups in the matadata file or in memory - so that you can use some column for an index - is a reasonable one that someone may still want to implement. That could happen in dask or outside.
